GitHub: use bitcoind v26.0 for CI

guggero opened the pull request:
Depends on https://github.com/lightningnetwork/lnd/pull/8555.

## Change Description

See if things break in the CI if we update to `bitcoind v26.0`.

calvinrzachman commented on .github/workflows/main.yml:
```diff
@@ -21,7 +21,7 @@ defaults:
 
 env:
-  BITCOIN_VERSION: "23.0"
+  BITCOIN_VERSION: "26.0"
```
The way we have [this job](https://github.com/lightninglabs/docker-bitcoin-core/blob/0132c0cf3974546a3f71d81c86ebe52bc5414abd/.github/workflows/build.yaml) setup, leads to images with tags like `bitcoin-core:26` and `bitcoin-core:26-alpine`. I think we may have to either drop the minor version here or re-tag and push images with a new tag that includes the minor version.

guggero replied:
Ah, good point. Updated.

guggero commented:
Okay, so the failed unit tests were due to a number of reasons:
 - `bitcoind v26.0` doesn't have support for Unix Sockets (`ipc://`) in ZMQ anymore. Broken by https://github.com/bitcoin/bitcoin/pull/22087, will be fixed by https://github.com/bitcoin/bitcoin/pull/27679
 - We had frequent TCP listening port collisions, at least when running the tests locally
 - We didn't clean up temporary directories for `bitcoind` sometimes

I added a couple of commits to address these issues.

yyforyongyu commented on scripts/install_bitcoind.sh:
```diff
@@ -2,9 +2,9 @@
 docker pull lightninglabs/bitcoin-core:$BITCOIND_VERSION
 CONTAINER_ID=$(docker create lightninglabs/bitcoin-core:$BITCOIND_VERSION)
-sudo docker cp $CONTAINER_ID:/opt/bitcoin-$BITCOIND_VERSION/bin/bitcoind /usr/local/bin/bitcoind
+sudo docker cp $CONTAINER_ID:/opt/bitcoin-$BITCOIND_VERSION.0/bin/bitcoind /usr/local/bin/bitcoind
```
I guess this assumes we gonna always build with major versions?

guggero replied:
Yeah, which I think is okay for our CI. We'll notice quite quickly if this becomes an issue.

yyforyongyu commented on chainntnfs/test_utils.go:
```diff
@@ -170,6 +171,17 @@ func NewMiner(t *testing.T, netParams *chaincfg.Params, extraArgs []string,
+	// have unique listening addresses across all packages in case things
+	// run parallel. That's why we use the parent process id to generate
+	// unique addresses.
+	parentProcess := os.Getppid()
```
Feel like this can just be combined with `lntest.NewMiner` with some modifications, guess we can clean it up in future PRs.

yyforyongyu commented on lnwallet/test/test_interface.go:
```diff
@@ -3234,72 +3206,9 @@ func runTests(t *testing.T, walletDriver *lnwallet.WalletDriver,
-				"-disablewallet",
-				"-zmqpubrawblock="+zmqBlockHost,
-				"-zmqpubrawtx="+zmqTxHost,
+			chainConn := chainntnfs.NewBitcoindBackend(
```
🎉

yyforyongyu reviewed (commented):
Thanks for the cleanup! No major comments, just need to merge the dependent PR first. And the unit tests need to be fixed.

Roasbeef reviewed (approved):
LGTM 🎨

guggero commented on chainntnfs/test_utils.go:
```diff
@@ -218,15 +216,14 @@ func NewBitcoindBackend(t *testing.T, minerAddr string, txindex,
 
-	rpcPort := rand.Intn(65536-1024) + 1024
-	zmqBlockHost := "ipc:///" + tempBitcoindDir + "/blocks.socket"
-	zmqTxHost := "ipc:///" + tempBitcoindDir + "/tx.socket"
```
We don't mention this in the user guide, only ever `tcp://`. That's why I think not many people even noticed the regression. There is no tracking issue, just a PR that aimed to fix the regression (https://github.com/bitcoin/bitcoin/pull/28020) that was closed in favor of another PR that fixed it in a more complete manner (https://github.com/bitcoin/bitcoin/pull/27679) but is farther from completion as it looks.

yyforyongyu commented:
Unit tests still failing. One of the itest failures has uncleaned shutdown that still occupying a port, werid
```
 unable to create server: listen tcp 127.0.0.1:6563: bind: address already in use 
```

guggero commented:
Yeah, it seems like I have to do some more digging into failures.

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

yyforyongyu commented on scripts/install_bitcoind.sh:
```diff
@@ -2,9 +2,9 @@
 set -ev
 
-BITCOIND_VERSION=${BITCOIN_VERSION:-25.0}
+BITCOIND_VERSION=${BITCOIN_VERSION:-25}
```
I think we should use one place to manage the version, to avoid mistakes such as 994a1426d6

ProofOfKeags replied:
Maybe this is a dumb question but why isn't this number 26?

guggero replied:
This is just the fallback version in case no environment variable is set. We set the version we actually want here: https://github.com/lightningnetwork/lnd/blob/ff3e8da1756572b0dd979a39b139a969498772dd/.github/workflows/main.yml#L24

But I guess we shouldn't do that to avoid confusion...

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

ProofOfKeags reviewed (approved):
I pulled this down and tested it and it looks like everything passes except a wtclient test (possibly a flake?). It doesn't seem related to this as best I can tell. I think that test will pass if you rebase this.

guggero commented:
@Roasbeef can you see the logs of the failed Travis build?
I only get this error message (not sure if we need to re-auth the app somehow?):

![Screenshot from 2024-02-02 18-37-28](https://github.com/lightningnetwork/lnd/assets/1008879/0f42e176-768e-45e6-b507-98ffacf5dafd)

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

lightninglabs-deploy commented:
@guggero, remember to re-request review from reviewers when ready

Roasbeef reviewed (changes requested):
This unit test looks to be failing across a few targets now: 
```
--- FAIL: TestFilteredChainView (34.34s)
    interface_test.go:871: Testing 'bitcoind_zmq' implementation of FilteredChainView
    interface_test.go:871: Testing 'bitcoind_polling' implementation of FilteredChainView
    --- FAIL: TestFilteredChainView/bitcoind_polling:_filtered_block_ntfns (20.21s)
        interface_test.go:225: filtered block notification didn't arrive
    interface_test.go:871: Testing 'p2p_neutrino' implementation of FilteredChainView
    interface_test.go:871: Testing 'btcd_websockets' implementation of FilteredChainView
FAIL
```

guggero commented:
Yeah, new failures keep popping up that I cannot easily reproduce locally. Or only with spending a lot of time on it, which I haven't found yet. Not sure if this is a priority, @saubyk?

saubyk commented:
> Yeah, new failures keep popping up that I cannot easily reproduce locally. Or only with spending a lot of time on it, which I haven't found yet. Not sure if this is a priority, @saubyk?

I am not aware of any critical dependency on bitcoind v26.0 for our imminent releases, so we can keep this at a lower priority

lightninglabs-deploy commented:
@bitromortac: review reminder
@ellemouton: review reminder
@crypt-iq: review reminder
@guggero, remember to re-request review from reviewers when ready

yyforyongyu commented:
Our CI is using bitcoind v23.0, and I think it's very important to test `lnd` against a newer version of `bitcoind`, there are new RPCs used but never checked in our CI, also the upcoming sweeper will also need a newer version of `bitcoind`.

Can take over if needed @guggero

guggero commented:
@yyforyongyu feel free to take over! I'm sure the failures are something small, just hard to find with some rounds of debugging required.

saubyk commented:
cc: @ProofOfKeags for taking this over

ProofOfKeags commented:
So here's what I've found out. The success/failure of the FilteredChainView tests is highly sensitive to whether other unit tests are running as well. This makes me think that our unit test suite somehow doesn't isolate the instances of the bitcoin node it is talking to.

- If I run *just* the FilteredChainView tests, it *always* succeeds.
- If I run the tests naively (allowing test caching), it usually succeeds in the limit case. Since only passing tests cache, I suspect that repeatedly running it stabilizes the test because fewer and fewer tests are competing for the same resource each run, reducing the race-y behavior
- If I run the tests with `nocache=1` it *always* fails.
- Extending the timeout on the FilteredChainView tests does not seem to have an impact. I increased it from 20s to 30s since the test seems to reliably run in 20s. A 50% boost should be enough if it was truly a time budget issue.

Taking into consideration that this broke by upgrading from Core 23 to Core 26 is more perplexing. I'd expect race-y behavior to not depend on that.

guggero commented:
Thank you @ProofOfKeags for investigating! Your insights have given me another idea what the problem might be.
The fact that running things in parallel causes the test to fail make me think the "per process unique port" generation mechanism doesn't work as expected. So I changed it to be unique per system, using a single file. Let's see how far that gets us...

guggero commented:
Okay, I think I have a version that should work. Wish me luck!

ProofOfKeags reviewed (commented):
So here's what I think is happening. What's going on in CI is identical afaict to what is happening for me locally.

There's a huge batch of errors related to "connection not found". Additionally, after all the tests have finished running on my machine if I inspect my Docker daemon, I find that there are some of these Postgres11 containers still running rather than being cleaned up. This leads me to believe quite strongly that there is a race condition in the resource allocator for the test instance of Postgres.

Verifying this further, I noticed that the function that is failing stating that it couldn't find the connection is the cleanup function that deallocates our resources. I bet that for some reason we are not atomically returning resources and registering deallocators.

I also noticed a possible concurrency issue wrt the filelocking behavior that will prevent intra-process concurrency from synchronizing properly.

Finally, I noticed that this is pinned to an old version of kvdb (1.2.2 + your mods).

ProofOfKeags commented on go.mod:
```diff
@@ -202,6 +200,8 @@ replace github.com/gogo/protobuf => github.com/gogo/protobuf v1.3.2
 // allows us to specify that as an option.
 replace google.golang.org/protobuf => github.com/lightninglabs/protobuf-go-hex-display v1.30.0-hex-display
 
+replace github.com/lightningnetwork/lnd/kvdb => github.com/guggero/lnd/kvdb v1.2.2-0.20240315183436-acfe249b2d09
```
This is making it very difficult for me to try and debug this further without having to reimplement your kvdb changes on top of 1.4.5

guggero replied:
If you look at the commit, that's just what's in #8555. I didn't push all tags to my fork, that's why the Golang pseudo version looks like 1.2.2.

guggero commented on kvdb/postgres/fixture.go:
```diff
@@ -8,86 +8,166 @@ import (
+}
 
-const testMaxConnections = 200
+var testPostgres *testPgFixture
```
This only is being used for the `make unit tags=kvdb_postgres` tests. In the normal `make unit` and `make unit-race` this shouldn't be used. So it could explain the failures in the Postgres test, but not in the others...

guggero commented on lntest/port/port.go:
```diff
@@ -0,0 +1,136 @@
+	for {
+		// Attempt to acquire the lock file. If it already exists, wait
+		// for a bit and retry.
+		lockFileHandle, err = os.OpenFile(
+			lockFile, os.O_CREATE|os.O_EXCL, 0600,
```
Hmm... Okay. I'll need to figure out a way to do this in a way that doesn't allow other threads to access the file...

ProofOfKeags replied:
Just also drop a global mutex in there. Mutex takes care of threads, lockfile takes care of processes.

guggero replied:
Okay, I added in the mutex, just to be on the safe side. But reading up on the documentation, it looks like the `O_EXCL` does what we expect here:

```
O_EXCL   int = syscall.O_EXCL   // used with O_CREATE, file must not exist.
```

The `OpenFile` function (or the underlying syscall) should be thread safe and therefore we should get an error if the file already exists at that point.

```
       O_EXCL Ensure that this call creates the file: if this flag is
              specified in conjunction with O_CREAT, and pathname
              already exists, then open() fails with the error EEXIST.
```
from https://man7.org/linux/man-pages/man2/open.2.html

guggero commented:
Okay, I fixed a couple of more issues:
 - Postgres: Needed to increase the number of allowed connections. Now the "connection not found" should be gone.
 - Travis: They seem to filter on the branch name and refuse to build because the word "bitcoin" appears in it. They suck anyway and GitHub now has a macOS build on M1 that we can use for our ARM (`arm64`) builds. So I replaced the Travis build with an additional GitHub run (let's see how that one fares).
 - The sporadic failures in the `routing/chainview` tests were caused by not waiting for the backend node to fully start up. There seems to be an issue with the `chain.BitcoindClient` that it can get weirdly de-synchronized if `Start()` and `NotifyBlocks()` are called while the node is syncing with the chain still (if things are very slow and blocks come in between `Start()` and `NotifyBlocks()`, weird things start to happen). Waiting one second before using a chain backend solved that issue for now.

Things look way better now IMO. There still are a couple of flakes, but I think those were all pre-existing.

guggero commented:
Okay, this is as green as it will probably ever get. There are still a couple of (pre-existing!) flakes in there. But I vote to get this merged since they very likely don't have anything to do with `bitcoind`.

yyforyongyu commented on .github/workflows/main.yml:
```diff
@@ -328,6 +328,44 @@ jobs:
           retention-days: 5
 
+  ########################
+  # run macOS integration test
```
niiice🎉

yyforyongyu reviewed (approved):
LGTM, thanks for the fix🙏

yyforyongyu commented:
btw can we update our merge rules to exclude the build from macOS?

ProofOfKeags commented on lntest/port/port.go:
```diff
@@ -0,0 +1,136 @@
+		select {
+		case <-timeout:
+			panic("timeout waiting for lock file")
+		case <-time.After(10 * time.Millisecond):
```
Ah I misunderstood what was happening here.

ProofOfKeags reviewed (approved):
🤝

guggero commented:
> btw can we update our merge rules to exclude the build from macOS?

It's currently not included in the list of required tests. So merges shouldn't be blocked on it (same for Windows).

pinheadmz commented:
FWIW https://github.com/bitcoin/bitcoin/pull/27679 is updated, has two ACKs and a review from an LND dev would be helpful to get a merge ;-)
